docs/tutorials/typescript/05-e2e-tests.md

```diff
@@ -0,0 +1,295 @@
+doc-type: [tutorial]
+doc-topic: [kubewarden, writing-policies, typescript, end-to-end-testing]
+doc-persona: [kubewarden-policy-developer]
+---
+
+<head>
+  <link rel="canonical" href="https://docs.kubewarden.io/tutorials/writing-policies/typescript/end-to-end-testing"/>
+</head>
+
+This section shows how you can write end-to-end tests running against the actual WebAssembly binary produced by the Javy compilation toolchain.
+
+## Prerequisites
+
+Recall, you need these tools on your development machine:
+
+- **bats**: Used to write the tests and automate their execution.
+- **kwctl**: CLI tool provided by Kubewarden to run its policies outside of Kubernetes, among other actions. It's covered in [this section](../../testing-policies/index.md) of the documentation.
+
+## Writing tests
+
+You'll be using bats to write and automate your tests. Each test has the following steps:
+
+1. Run the policy using `kwctl`.
+2. Perform assertions against the output produced by `kwctl`.
+
+All the end-to-end tests go in a file called `e2e.bats`. The project scaffolding project already includes an example `e2e.bats`. You need to extend its contents to provide comprehensive test coverage for your policy behavior.
+
+## Test data files
+
+The end-to-end tests use JSON files containing Kubernetes resources. You should have these test data files:
+
+**`test_data/pod.json`** - A Pod without a hostname:
+
+```json
+{
+  "apiVersion": "v1",
+  "kind": "Pod",
+  "metadata": {
+    "name": "test-pod"
+  },
+  "spec": {
+    "containers": [
+      {
+        "name": "nginx",
+        "image": "nginx:latest"
+      }
+    ]
+  }
+}
+```
+
+**`test_data/pod_with_hostname.json`** - A Pod with a hostname:
+
+```json
+{
+  "apiVersion": "v1",
+  "kind": "Pod",
+  "metadata": {
+    "name": "test-pod"
+  },
+  "spec": {
+    "hostname": "test-hostname",
+    "containers": [
+      {
+        "name": "nginx",
+        "image": "nginx:latest"
+      }
+    ]
+  }
+}
+```
+
+**`test_data/deployment.json`** - A non-Pod resource:
+
+```json
+{
+  "apiVersion": "apps/v1",
+  "kind": "Deployment",
+  "metadata": {
+    "name": "test-deployment"
+  },
+  "spec": {
+    "replicas": 1,
+    "selector": {
+      "matchLabels": {
+        "app": "test"
+      }
+    },
+    "template": {
+      "metadata": {
+        "labels": {
+          "app": "test"
+        }
+      },
+      "spec": {
+        "containers": [
+          {
+            "name": "nginx",
+            "image": "nginx:latest"
+          }
+        ]
+      }
+    }
+  }
+}
+```
+
+## Basic test cases
+
+The template already provides several basic tests. Here are the existing tests with explanations:
+
+### Test 1: Reject denied hostnames
+
+```bash
+@test "reject because hostname is on deny list" {
+  run kwctl run annotated-policy.wasm -r test_data/pod_with_hostname.json --settings-json '{"denied_hostnames": ["forbidden-host", "test-hostname"]}'
```

Comment:
Are you sure this works? `kwctl run` takes a parameter a Kubernetes `AdmissionRequest` object via its `-r` flag. What you're passing is a Kubernetes object definition.

You should either use a fixture that has an `AdmissionRequest` defined inside of it, or you should use `kwctl scaffold admission-request ...` to convert that Pod definition into an `AdmissionRequest` to be fed into `kwctl run`.